internal/aws/ssm_remote_access_strategy.go

```diff
@@ -146,64 +245,144 @@ func (s *SSMRemoteAccessStrategy) CleanupSSMManagedNodes(ctx context.Context, po
 	return nil
 }
 
-// isSSMRegistrationCompleted checks if SSM registration is already done for this pod
-func (s *SSMRemoteAccessStrategy) isSSMRegistrationCompleted(ctx context.Context, pod *corev1.Pod) bool {
+// isContainerRunning checks if a container is in running state
+func (s *SSMRemoteAccessStrategy) isContainerRunning(pod *corev1.Pod, containerName string) bool {
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.Name == containerName {
+			return cs.State.Running != nil
+		}
+	}
+	return false
+}
+
+// getCurrentRestartCounts extracts restart counts from pod status
+func (s *SSMRemoteAccessStrategy) getCurrentRestartCounts(pod *corev1.Pod) (sidecarCount, workspaceCount int32) {
+	for _, cs := range pod.Status.ContainerStatuses {
+		if cs.Name == SSMAgentSidecarContainerName {
+			sidecarCount = cs.RestartCount
+		}
+		if cs.Name == WorkspaceContainerName {
+			workspaceCount = cs.RestartCount
+		}
+	}
+	return sidecarCount, workspaceCount
+}
+
+// readRegistrationState reads state from shared volume
+// Returns:
+//   - (*RegistrationState, nil) if state file exists and is valid
+//   - (nil, nil) if state file doesn't exist (first time setup)
+//   - (nil, error) if state file exists but is corrupted
+func (s *SSMRemoteAccessStrategy) readRegistrationState(ctx context.Context, pod *corev1.Pod) (*RegistrationState, error) {
```

Comment:
as we discussed offline, we need to think about better way to address the race condition especially the side car container case. The ssm related status(de-register, resgister) could be tricky when race condition happens. 

Given it's overall improving the UX, this could be p0.5 and not blocking.